sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -173,7 +194,27 @@ class GpuSortAggregateMeta(
         case "partial" => if (hashAggMode.contains(Final) || hashAggMode.contains(Complete)) {
           // replacing only Partial hash aggregates, so a Final or Commplete one should not replace
           willNotWorkOnGpu("Replacing Final or Complete hash aggregates disabled")
-         }
+        }
+          // In partial mode, if there are non distinct functions and multiple distinct functions
+          // non distinct functions are computed using First operator. The final result would be
+          // incorrect for non distinct functions for partition size > 1. Reason for this is - if
+          // the first batch computed and sent to CPU doesn't contain all the rows required to
+          // compute non distinct function(s), then Spark would consider that value as final result
+          // (due to First)Falling back to CPU for this special case.
+          if (agg.aggregateExpressions.exists(e => e.aggregateFunction.isInstanceOf[First])) {
+            val distinctAggs = agg.aggregateExpressions.flatMap(expr =>
+              expr.children.map {
+                _.collect {
```

Comment:
I'll run this code in debug to check more, but if you already know the answer, please do elaborate, how do we figure out that the counts and the averages being counted are distinct. Also, doesn't this case fall back when there are say 1 count distinct and 1 avg distinct for example. Is that a valid case for the plan to fall back ?

Reply:
As for the collect, why can't we just count the occurrences instead of creating lists just for the size

Reply:
Sure, When we look at the physical plan, `partial_first` is included in the plan if there are multiple distinct aggregation functions. So, here I am checking if in the partial mode and  the aggregregateFunction is  instance of `First`. If these conditions are true, then  we are counting number of distinct aggregatation in that AggregateExpressions.
If we have 1 count distinct and 1 avg distinct it has to fall back to CPU. Number of distinct Aggs is what is considered.

https://github.com/apache/spark/blob/1597d8fcd4c68e723eb3152335298c7d05155643/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala#L202

Reply:
Thanks! Good to know that avg distinct +count distinct falls in the same area. Maybe we can add it to the test as well is what i was getting at.

As for the first, I understand we are checking for first and then looking at agg expressions, how do we know that they are distinct and not non-distincts? Also, can we remove the collect and count the number of distinct aggs.

Reply:
Looking at the spark code some more, sum distinct and others might need to be tested/added as well..

Reply:
As for 
> . I ran timings over this loop and runtime was negligible. 
>
Nit. While such tests can be useful sometimes, it is not a sufficient test to prove that collect is efficient or not.

Reply:
Also, `partial on gpu: empty df: float basic aggregates group by floats` fails with this patch. Like suspected, it is fsalling back for non distinct cases as well.

Reply:
As for counting : Here is an example, feel free to improve on it, javent tested it but it is cleaner to me.
```          
        if (agg.aggregateExpressions.exists(e => e.aggregateFunction.isInstanceOf[First])) {
          val count = agg.aggregateExpressions.flatMap(expr => expr.children).count {
            expr => expr match {
            case a: Count => true
            case b: Average => true
          }
          }
          if (count > 1) {
            willNotWorkOnGpu("Aggregate of non distinct functions with multiple distinct " +
              "functions is non deterministic for non distinct functions as it is computed using " +
              "First.")
          }
        }
```